src/packagedcode/opam.py

```diff
@@ -0,0 +1,309 @@
+import io
+import logging
+import re
+
+import attr
+from packageurl import PackageURL
+
+from commoncode import filetype
+from commoncode import fileutils
+from packagedcode import models
+
+
+"""
+Handle opam package.
+"""
+
+TRACE = False
+
+logger = logging.getLogger(__name__)
+
+if TRACE:
+    import sys
+    logging.basicConfig(stream=sys.stdout)
+    logger.setLevel(logging.DEBUG)
+
+
+@attr.s()
+class OpamPackage(models.Package):
+    metafiles = ('*opam',)
+    extensions = ('.opam',)
+    default_type = 'opam'
+    default_primary_language = 'Ocaml'
+    default_web_baseurl = 'https://opam.ocaml.org/packages'
+    default_download_baseurl = None
+    default_api_baseurl = None
+
+    @classmethod
+    def recognize(cls, location):
+        yield parse(location)
+
+    @classmethod
+    def get_package_root(cls, manifest_resource, codebase):
+        return manifest_resource.parent(codebase)
+
+    def repository_homepage_url(self, baseurl=default_web_baseurl):
+        if self.name:
+            return '{}/{}'.format(baseurl, self.name)
+
+
+def is_opam(location):
+    if location.endswith('opam'):
+        return True
+
+
+def parse(location):
+    """
+    Return a Package object from a opam file or None.
+    """
+    if not is_opam(location):
+        return
+
+    package_data = parse_opam(location)
+    return build_opam_package(package_data)
+
+
+def build_opam_package(package_data):
+    """
+    Return a Package from a opam file or None.
+    """
+    package_dependencies = []
+    deps = package_data.get('depends') or []
+    for dep in deps:
+        package_dependencies.append(
+            models.DependentPackage(
+                purl=dep.purl,
+                requirement=dep.version,
+                scope='dependency',
+                is_runtime=True,
+                is_optional=False,
+                is_resolved=False,
+            )
+        )
+
+    name = package_data.get('name')
+    maintainers = package_data.get('maintainer') or []
+    authors = package_data.get('authors') or []
+    homepage_url = package_data.get('homepage')
+    vcs_url = package_data.get('dev-repo')
+    summary = package_data.get('synopsis')
+    description = package_data.get('description')
+    if summary:
+        description = summary
+    elif summary and description:
+        if len(summary) > len(description):
+            description = summary
+    
+    parties = []
+    for author in authors:
+        parties.append(
+            models.Party(
+                type=models.party_person,
+                name=author,
+                email=None,
+                role='author'
+            )
+        )
+    
+    for maintainer in maintainers:
+        parties.append(
+            models.Party(
+                type=models.party_person,
+                name=None,
+                email=maintainer,
+                role='email'
+            )
+        )
+
+    package = OpamPackage(
+        name=name or None,
```

Comment:
Do you need or None here? and is it ever possible to have something empty?

Reply:
In some opam files, name is not mentioned. I have given some examples in sample test files.